README.md
# WinSystems-NVRAM_SSD_Linux_Block_Device_Driver

A linux block device driver for the non-volatile SRAM "SSD" on some of WinSystems platforms

1. Install Source Code

Build Instructions:



documentation/Release_Notes.txt

Release notes for the NV-SRAM SSD Linux device driver

   Date        Revision      Notes
----------    ----------    -------------------------------------------------------
07/01/2011       v1.0       Original release, for kernel v 2.6.32
12/11/2019       v1.01      Updated for kernel v 4.18.0

Tests/test1.sh
#!/bin/bash
#
# Shell script definitions for testing the SSD driver
#
. include.sh

rm -f $RESULTS_FILE

#
# test for total zeroing
#
dd if=$ZERO_1MB_FILE of=$SSD_DEVICE
dd if=$SSD_DEVICE of=$CAPTURE_FILE
echo "Test of complete device zeroing" >> $RESULTS_FILE
diff $CAPTURE_FILE $ZERO_1MB_FILE >> $RESULTS_FILE

###########################################################################################
#
# test for complete fill with pre-generated random data
#
dd if=$RANDOM_1MB_FILE of=$SSD_DEVICE
dd if=$SSD_DEVICE of=$CAPTURE_FILE
echo "Test of complete random fill" >> $RESULTS_FILE
diff $CAPTURE_FILE $RANDOM_1MB_FILE >> $RESULTS_FILE

#
# zero array
#
dd if=$ZERO_1MB_FILE of=$SSD_DEVICE

#
# create 1 sector compare file from 1st 512 bytes of random file
#
dd if=$RANDOM_1MB_FILE of=$COMPARE_FILE bs=$SECTOR_SIZE count=1

#
# fill each sector with the same random block and compare
#
for ((sector=0;sector<SECTOR_LAST;sector++)); do
   dd if=$RANDOM_1MB_FILE of=$SSD_DEVICE bs=$SECTOR_SIZE count=1 skip=0 seek=$((sector))
   dd if=$SSD_DEVICE of=$CAPTURE_FILE bs=$SECTOR_SIZE count=1 skip=$((sector)) 
   echo "Compare sector = $sector" >> $RESULTS_FILE
   diff $COMPARE_FILE $CAPTURE_FILE >> $RESULTS_FILE
done

#
# zero array
#
dd if=$ZERO_1MB_FILE of=$SSD_DEVICE

#
# zero array, create compare file
#
dd if=$ZERO_1MB_FILE of=$SSD_DEVICE
dd if=$RANDOM_1MB_FILE of=$COMPARE_FILE bs=$SECTOR_SIZE count=1

#
# offset 1/2 into each sector, fill with 512 bytes of the same random block, read back and compare
#
for ((sector=0;sector<(SECTOR_LAST-1);sector++)); do
   dd if=$RANDOM_1MB_FILE of=$SSD_DEVICE bs=$SECTOR_SIZE count=1 skip=0 seek=$(((sector*SECTOR_SIZE) + 256))
   dd if=$SSD_DEVICE of=$CAPTURE_FILE bs=$SECTOR_SIZE count=1 skip=$(((sector*SECTOR_SIZE) + 256 )) seek=0
   echo "Compare with offset = 256, sector = $sector" >> $RESULTS_FILE
   diff $COMPARE_FILE $CAPTURE_FILE >> $RESULTS_FILE
done

#
# fill entire array 1 byte at a time from pre-generated random data
#
for ((offset=0;offset<SSD_SIZE_IN_BYTES;offset++)); do
   dd if=$RANDOM_1MB_FILE of=$SSD_DEVICE bs=1 count=1 skip=$((offset)) seek=$((offset))
done

echo "Test of individual byte access ( block size=1, count = $SSD_SIZE_IN_BYTES )" >> $RESULTS_FILE
dd if=$SSD_DEVICE of=$CAPTURE_FILE bs=$SECTOR_SIZE count=$SECTOR_COUNT skip=0
diff $CAPTURE_FILE $RANDOM_1MB_FILE >> $RESULTS_FILE




Tests/include.sh
#!/bin/bash
#
# Shell script definitions for testing the SSD driver
#

RESULTS_FILE=results.txt
RESULTS_ODD_FILE=results_odd.txt
RESULTS_EVEN_FILE=results_even.txt

SSD_DEVICE=/dev/ssd
ZERO_DEVICE=/dev/zero

RANDOM_1MB_FILE=random_1MB.raw
RANDOM_4096_FILE=random_4096.raw

ZERO_1MB_FILE=zero_1MB.raw
ZERO_2048_FILE=zero_2048.raw
ZERO_512_FILE=zero_512.raw

CAPTURE_FILE=capture.raw
CAPTURE_ODD_FILE=capture_odd.raw
CAPTURE_EVEN_FILE=capture_even.raw

COMPARE_FILE=compare.raw
COMPARE_ODD_FILE=compare_odd.raw
COMPARE_EVEN_FILE=compare_even.raw

SSD_SIZE_IN_BYTES=1048576
SECTOR_SIZE=512

SECTOR_OFFSET_MAX=511
SECTOR_COUNT=2048
SECTOR_LAST=2047

BYTE_SEQ_FILL_END=255


Tests/background_odd.sh
#!/bin/bash
#
# Shell script definitions for testing the SSD driver
#
. include.sh

rm -f $RESULTS_ODD_FILE

dd if=$RANDOM_1MB_FILE of=$COMPARE_ODD_FILE bs=$SECTOR_SIZE count=1

#
# fill each sector with the same random block and compare
#
for ((sector=0;sector<1023;sector++)); do
   dd if=$RANDOM_1MB_FILE of=$SSD_DEVICE bs=$SECTOR_SIZE count=1 skip=0 seek=$((sector))
   dd if=$SSD_DEVICE of=$CAPTURE_ODD_FILE bs=$SECTOR_SIZE count=1 skip=$((sector)) 
   echo "Compare sector = $sector" >> $RESULTS_ODD_FILE
   diff $COMPARE_ODD_FILE $CAPTURE_ODD_FILE >> $RESULTS_ODD_FILE
done


ssd.h
//*****************************************************************************
//
//   Copyright 2019, WinSystems Inc.
//
//   Permission is hereby granted, free of charge, to any person obtaining a 
//   copy of this software and associated documentation files (the "Software"), 
//   to deal in the Software without restriction, including without limitation 
//   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
//   and/or sell copies of the Software, and to permit persons to whom the 
//   Software is furnished to do so, subject to the following conditions:
//
//   The above copyright notice and this permission notice shall be included 
//   in all copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
//   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
//   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
//   DEALINGS IN THE SOFTWARE.
//
//*****************************************************************************
//
//    Name       : ssd.h
//
//    Project    : NV-RAM SSD For WinSystems platforms
//
//    Author     : Paul DeMetrotion
//
//    Description:
//             This module contains the definitions for the NV-RAM SSD
//             block device driver.
//          
//
//*****************************************************************************
//
//      Date         Revision    Change Description
//    --------       --------    ---------------------------------------------
//    2019/12/11       1.01       Modified to support kernel 4.18
//
//*****************************************************************************
#ifndef CHARDEV_H
  #define CHARDEV_H

#include <linux/ioctl.h>

#define SSD_SIZE_MB 1
#define SSD_MINORS 4
#define NSECTORS 2048
#define LOGICAL_BLOCK_SIZE 512
#define KERNEL_SECTOR_SIZE 512
#define IOCTL_NUM 'e'
#define SUCCESS 0

// enable write protect
#define IOCTL_EN_WP _IOWR(IOCTL_NUM, 1, int)

// disable write protect
#define IOCTL_DIS_WP _IOWR(IOCTL_NUM, 2, int)

// disable write protect
#define IOCTL_DMP_SECT _IOWR(IOCTL_NUM, 3, unsigned long)

#endif


lock.c
//*****************************************************************************
//
//   Copyright 2019, WinSystems Inc.
//
//   Permission is hereby granted, free of charge, to any person obtaining a 
//   copy of this software and associated documentation files (the "Software"), 
//   to deal in the Software without restriction, including without limitation 
//   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
//   and/or sell copies of the Software, and to permit persons to whom the 
//   Software is furnished to do so, subject to the following conditions:
//
//   The above copyright notice and this permission notice shall be included 
//   in all copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
//   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
//   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
//   DEALINGS IN THE SOFTWARE.
//
//*****************************************************************************
//
//    Name       : lock.c
//
//    Project    : NVRAM SSD Device Driver
//
//    Author     : Paul DeMetrotion
//
//    Description:
//             This module contains the implementation lock/unlock application
//             for the block device driver for the NV-RAM SSD
//
//*****************************************************************************
//
//      Date         Revision    Change Description
//    --------       --------    ---------------------------------------------
//    2019/12/11      1.01      Updated for kernel v4.18
//
//*****************************************************************************
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>

// Include the WinSystems UIO48 definitions
#include "ssd.h"

int main(int argc, char *argv[])
{
	// arguments for on or off
	if (argc != 2)
	{
		printf("Usage: lock <on/off>\n");
		exit(1);
	}

	if (argv[1][1] == 0x6E)	// 'n'
		if(ssd_set_wp() < 0) 
		{
			fprintf(stderr, "Can't access device SSD - Aborting\n");
			exit(1);
		}
		else
			printf("SSD write protect enabled\n");
	else
		if(ssd_clr_wp() < 0)
		{
			fprintf(stderr, "Can't access device SSD - Aborting\n");
			exit(1);
		}
		else
			printf("SSD write protect disabled\n");

	return SUCCESS;
}


ssdio.c
//*****************************************************************************
//
//   Copyright 2019, WinSystems Inc.
//
//   Permission is hereby granted, free of charge, to any person obtaining a 
//   copy of this software and associated documentation files (the "Software"), 
//   to deal in the Software without restriction, including without limitation 
//   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
//   and/or sell copies of the Software, and to permit persons to whom the 
//   Software is furnished to do so, subject to the following conditions:
//
//   The above copyright notice and this permission notice shall be included 
//   in all copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
//   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
//   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
//   DEALINGS IN THE SOFTWARE.
//
//*****************************************************************************
//
//    Name       : ssdio.c
//
//    Project    : NVRAM SSD library
//
//    Author     : Paul DeMetrotion
//
//    Description:
//             This module contains the implementation of the block device
//             driver library for the NV-RAM SSD
//          
//
//*****************************************************************************
//
//      Date         Revision    Change Description
//    --------       --------    ---------------------------------------------
//    2019/12/11      1.01       Updated for kernel v4.18
//
//*****************************************************************************

#include <stdio.h>
#include <fcntl.h>      /* open */ 
#include <unistd.h>     /* exit */
#include <sys/ioctl.h>  /* ioctl */

#include "ssd.h"    

//**************************************************************************
//		USER LIBRARY FUNCTIONS
//**************************************************************************

// device handles
int handle = 0;

// the names of our device nodes
char *device_id = "/dev/ssd";

//
//------------------------------------------------------------------------
//
// ssd_set_wp - enables write protect
//
// Arguments:
//
// Returns:
//
//------------------------------------------------------------------------
//
int ssd_set_wp(void)
{
    if(check_handle())   // check for chip availability
		return -1;

	return ioctl(handle, IOCTL_EN_WP, 0);
}

//
//------------------------------------------------------------------------
//
// ssd_clr_wp - enables write protect
//
// Arguments:
//
// Returns:
//
//------------------------------------------------------------------------
//
int ssd_clr_wp(void)
{
    if(check_handle())   // check for chip availability
		return -1;

	return ioctl(handle, IOCTL_DIS_WP, 0);
}

//
//------------------------------------------------------------------------
//
// ssd_set_wp - enables write protect
//
// Arguments:
//
// Returns:
//
//------------------------------------------------------------------------
//
int ssd_dmp_sect(int sect)
{
    if(check_handle())   // check for chip availability
		return -1;

	return ioctl(handle, IOCTL_DMP_SECT, sect * 512);
}

//
//------------------------------------------------------------------------
//
// check_handle - Checks that a handle to the device file exists.
//				  If it does not a file open is performed.
//
// Arguments:
//
// Returns:
//			0		Handle is valid
//			-1		Chip does not exist or it's handle is invalid
//
//------------------------------------------------------------------------
//
int check_handle(void)
{
    if(handle > 0)	// If it's already a valid handle
		return 0;

    if(handle == -1)	// If it's already been tried
		return -1;

	// Try opening the device file, in case it hasn't been opened yet
    handle = open(device_id, O_RDWR);

    if(handle > 0)	// If it's now a validopen handle
		return 0;
    
    handle = -1;
		return -1;
}


ssd.c
//*****************************************************************************
//
//   Copyright 2019, WinSystems Inc.
//
//   Permission is hereby granted, free of charge, to any person obtaining a 
//   copy of this software and associated documentation files (the "Software"), 
//   to deal in the Software without restriction, including without limitation 
//   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
//   and/or sell copies of the Software, and to permit persons to whom the 
//   Software is furnished to do so, subject to the following conditions:
//
//   The above copyright notice and this permission notice shall be included 
//   in all copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
//   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
//   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
//   DEALINGS IN THE SOFTWARE.
//
//*****************************************************************************
//
//    Name       : ssd.c
//
//    Project    : NVRAM SSD Device Driver
//
//    Author     : Paul DeMetrotion
//
//    Description:
//             This module contains the implementation of the block device
//             driver for the NV-RAM SSD
//          
//
//*****************************************************************************
//
//      Date         Revision    Change Description
//    --------       --------    ---------------------------------------------
//    2019/12/11      1.01      Updated for kernel v4.18
//
//*****************************************************************************

static char *RCSInfo = "$Id: ssd.c,v 1.01 2019/11/28 20:31:10 Paul DeMetrotion $";

#ifndef __KERNEL__
	#define __KERNEL__
#endif

#ifndef MODULE
	#define MODULE
#endif

#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/init.h>
#include <linux/kernel.h>
#include <linux/fs.h>
#include <linux/errno.h>
#include <linux/types.h>
#include <linux/vmalloc.h>
#include <linux/genhd.h>
#include <linux/blkdev.h>
#include <linux/hdreg.h>
#include <linux/version.h>
#include <asm/io.h>
#include "ssd.h"

#define DRVR_NAME		"ssd"
#define DRVR_VERSION	"1.0"
#define DRVR_RELDATE	"29Jun2011"

//#define DEBUG 1

// Function prototypes for local functions 

// ssd block device structure
static struct ssd_bdevice {
	int size;
	short users;
	spinlock_t lock;
	struct gendisk *gd;
	int wp_flag;
} ssd_bdev;

// request queue
static struct request_queue *queue;

// Driver major number
static int ssd_init_major = 0;	// 0 = allocate dynamically
static int ssd_major;

// Our modprobe command line arguments
static int io = 0;

///**********************************************************************
//			DEVICE SUBROUTINES
///**********************************************************************
static void ssd_read(unsigned long offset, char *buffer)
{
	int i, ptr;

	#ifdef DEBUG
	printk ("<1>SSD - ssd_read\n");
	printk ("<1> Starting Address = 0x%06lX\n", offset);
	#endif
		
	for (i = 0; i < 2; i++)
	{
		outb(offset >> 16, io);
		outb(offset >> 8, io + 1);

		for (ptr = 0; ptr < 256; ptr++)
			*buffer++ = inb(io + 2);

		offset += 256;
	}

	#ifdef DEBUG
	printk ("<1>SSD - ssd_read complete\n");
	#endif
}

static void ssd_write(unsigned long offset, char *buffer)
{
	int i, ptr;

	#ifdef DEBUG
	printk ("<1>SSD - ssd_write\n");
	printk ("<1> Starting Address = 0x%06lX\n", offset);
	#endif
		
	for (i = 0; i < 2; i++)
	{
		outb(offset >> 16, io);
		outb(offset >> 8, io + 1);

		for (ptr = 0; ptr < 256; ptr++)
			outb(*buffer++, io + 2);

		offset += 256;
	}

	#ifdef DEBUG
	printk ("<1>SSD - ssd_write complete\n");
	#endif
}

static void ssd_transfer(struct ssd_bdevice *bdev, sector_t sector, unsigned long nsect, char *buffer, int write)
{
	unsigned long offset = sector * LOGICAL_BLOCK_SIZE;
	unsigned long nbytes = nsect * LOGICAL_BLOCK_SIZE;

	if ((offset + nbytes) > bdev->size)
	{
		#ifdef DEBUG
		printk ("<1>SSD - Beyond-end write (0x%06lX %ld)\n", offset, nbytes);
		#endif
		
		return;
	}
	
	if (write && !(bdev->wp_flag))
	{
		while (nsect--)
		{
			ssd_write(offset, buffer);
			offset += LOGICAL_BLOCK_SIZE;
			buffer += LOGICAL_BLOCK_SIZE;
		}
	}
	else if (write && bdev->wp_flag)
	{
		printk ("<1>SSD - Write Protect mode is enabled. To disable run the lock program.\n");
		
		#ifdef DEBUG
		printk ("<1>SSD - Protected write (0x%06lX %ld)\n", offset, nbytes);
		#endif
	}
	else
	{
		while (nsect--)
		{
			ssd_read(offset, buffer);
			offset += LOGICAL_BLOCK_SIZE;
			buffer += LOGICAL_BLOCK_SIZE;
		}
	}
}

static void ssd_request(struct request_queue *q)
{
	struct request *req;

	req = blk_fetch_request(q);

	while (req != NULL)
	{
		char *whereto;

#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,10,0)
		if (req == NULL || (req->cmd_type != REQ_TYPE_FS))
#else
		if(req == NULL) 
#endif
		{
			#ifdef DEBUG
			printk ("<1>SSD - Skip non-CMD request\n");
			#endif
			__blk_end_request_all(req, -EIO);
			continue;
		}
	
#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,10,0)
		whereto = req->buffer;
#else
		whereto = bio_data(req->bio);
#endif
		ssd_transfer(&ssd_bdev, blk_rq_pos(req), blk_rq_cur_sectors(req), whereto, rq_data_dir(req));
 
		if (!__blk_end_request_cur(req, 0))
			req = blk_fetch_request(q);
	}
}

///**********************************************************************
//			DEVICE OPEN
///**********************************************************************
static int ssd_open(struct block_device *bdev, fmode_t mode)
{
	ssd_bdev.users++;

	#ifdef DEBUG
	printk ("<1>SSD - ssd_open %d\n", ssd_bdev.users);
	#endif

	return SUCCESS;
}

///**********************************************************************
//			DEVICE CLOSE
///**********************************************************************
#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,10,0)
static int ssd_release(struct gendisk *gd, fmode_t mode)
#else
static void ssd_release(struct gendisk *gd, fmode_t mode)
#endif

{
	#ifdef DEBUG
	printk ("<1>SSD - ssd_release %d\n", ssd_bdev.users);
	#endif 

	ssd_bdev.users--;


#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,10,0)
	return SUCCESS;
#endif
}

///**********************************************************************
//			DEVICE IOCTL
///**********************************************************************
int ssd_ioctl(struct block_device *bdev, fmode_t mode, unsigned int ioctl_num, unsigned long ioctl_param)
{
	#ifdef DEBUG
	int i;

	printk("<1>SSD - ssd_ioctl(%d)\n", ioctl_num);
	#endif

	// Switch according to the ioctl called
	switch (ioctl_num) {
		case IOCTL_EN_WP:
			outb(0, io + 4);	// enable write protect
			ssd_bdev.wp_flag = 1;			// structure flag
			return SUCCESS;

		case IOCTL_DIS_WP:
			outb(1, io + 4);	// disable write protect
			ssd_bdev.wp_flag = 0;		// structure flag
			return SUCCESS;

		#ifdef DEBUG
		case IOCTL_DMP_SECT:
			// set address
			outb(ioctl_param >> 16, io);
			outb(ioctl_param >> 8, io + 1);

			// read & display sector
			for (i=0; i < 512; i++)
			{
				if (i % 16 == 0) printk("%05lX : ", ioctl_param + i);
				printk("%02X ", inb(io + 2));
				if (i % 16 == 15) printk("\n");
			}
	
			return SUCCESS;
		#endif

		// Catch all return
		default:
			return(-EINVAL);
	}
}

///**********************************************************************
//			DEVICE GETGEO
///**********************************************************************
int ssd_getgeo(struct block_device *bdev, struct hd_geometry *geo)
{
	long size;

	#ifdef DEBUG
	printk ("<1>SSD - ssd_getgeo\n");
	#endif

	/* We have no real geometry, of course, so make something up. */
	size = ssd_bdev.size * (LOGICAL_BLOCK_SIZE / KERNEL_SECTOR_SIZE);
	geo->cylinders = (size & ~0x3f) >> 6;
	geo->heads = 4;
	geo->sectors = 16;
	geo->start = 0;
	
	return SUCCESS;
}

///**********************************************************************
//			Module Declarations
// This structure will hold the functions to be called 
// when a process does something to the device
///**********************************************************************
static struct block_device_operations ssd_fops = {
	.open			=	ssd_open,
	.release		=	ssd_release,
	.ioctl			=	ssd_ioctl,
	.getgeo			=	ssd_getgeo,		// need for fdisk support
	.owner			=	THIS_MODULE
};

///**********************************************************************
//			MODULE INITIALIZATION
///**********************************************************************
static int __init ssd_init(void)
{
	int ret_val;

	// Sign-on
	printk("<1>WinSystems, Inc. SSD Linux Device Driver\n");
	printk("<1>Copyright 2011, All rights reserved\n");
	printk("<1>%s\n", RCSInfo);

	// check and map our I/O region requests
	if(request_region(io, 8, DRVR_NAME) == NULL)
	{
		printk("<1>SSD - Unable to use I/O Address %4Xh\n", io);
		return -ENODEV;
	}
	else
		printk("<1>SSD - Base I/O Address = %4Xh\n", io);
		
	// register the block device
	ssd_major = register_blkdev(ssd_init_major, DRVR_NAME);

	if(ssd_major < 0)
	{
		printk("<1>SSD - Cannot obtain major number\n");
		ret_val = ssd_major;
		goto exit_release_io;
	}
	else
	{
		if (ssd_init_major)
			ssd_major = ssd_init_major;

		printk("<1>SSD - Major number %d assigned\n", ssd_major);
	}

	// initialize structure
	memset(&ssd_bdev, 0, sizeof(struct ssd_bdevice));
	ssd_bdev.size = NSECTORS * LOGICAL_BLOCK_SIZE;

	// spin lock
	spin_lock_init(&ssd_bdev.lock);

	// request queue
	queue = blk_init_queue(ssd_request, &ssd_bdev.lock);

	if (queue == NULL)
	{
		printk("<1>SSD - No memory resources for request queue\n");
		ret_val = -ENOMEM;
		goto exit_bdev_unregister;
	}

	// gendisk
	ssd_bdev.gd = alloc_disk(SSD_MINORS);

	if (ssd_bdev.gd == NULL)
	{
		printk("<1>SSD - No memory resources for gendisk structure\n");
		ret_val = -ENOMEM;
		goto exit_reqq_delete;
	}

	// initialize gendisk
	ssd_bdev.gd->major = ssd_major;
	ssd_bdev.gd->first_minor = 0;
	ssd_bdev.gd->fops = &ssd_fops;
	ssd_bdev.gd->queue = queue;
	ssd_bdev.gd->private_data = &ssd_bdev;
	strcpy(ssd_bdev.gd->disk_name, DRVR_NAME);
	set_capacity(ssd_bdev.gd, NSECTORS * (LOGICAL_BLOCK_SIZE / KERNEL_SECTOR_SIZE));

	// initialize wp flag to off
	ssd_bdev.wp_flag = 1;

	// add disk at end of init
	add_disk(ssd_bdev.gd);

	return SUCCESS;

exit_reqq_delete:
	blk_cleanup_queue(queue);

exit_bdev_unregister:
	unregister_blkdev(ssd_major, DRVR_NAME);
	ssd_major = 0;

exit_release_io:
	release_region(io, 8);

	return ret_val;
}

///**********************************************************************
//			MODULE EXIT
///**********************************************************************
static void ssd_exit(void)
{
	// delete gendisk
	del_gendisk(ssd_bdev.gd);

	// delete request queue
	blk_cleanup_queue(queue);

	// unregister the device
	unregister_blkdev(ssd_major, DRVR_NAME);
	ssd_major = 0;

	// release io address range
	release_region(io, 8);
}  

module_param(io, int, S_IRUGO);

module_init(ssd_init);
module_exit(ssd_exit);

MODULE_LICENSE("MIT");
MODULE_DESCRIPTION("WinSystems,Inc. SSD Device Driver");
MODULE_AUTHOR("Paul DeMetrotion");
